zerver/models/custom_profile_fields.py

```diff
@@ -29,29 +29,34 @@
     validate_select_field,
 )
 from zerver.models.realms import Realm
-from zerver.models.users import UserProfile, get_user_profile_by_id_in_realm
+from zerver.models.users import UserProfile
 
 
 def check_valid_user_ids(realm_id: int, val: object, allow_deactivated: bool = False) -> List[int]:
     user_ids = check_list(check_int)("User IDs", val)
-    realm = Realm.objects.get(id=realm_id)
-    for user_id in user_ids:
-        # TODO: Structurally, we should be doing a bulk fetch query to
-        # get the users here, not doing these in a loop.  But because
-        # this is a rarely used feature and likely to never have more
-        # than a handful of users, it's probably mostly OK.
-        try:
-            user_profile = get_user_profile_by_id_in_realm(user_id, realm)
-        except UserProfile.DoesNotExist:
-            raise ValidationError(_("Invalid user ID: {user_id}").format(user_id=user_id))
-
-        if not allow_deactivated and not user_profile.is_active:
+    user_profiles = UserProfile.objects.filter(realm_id=realm_id, id__in=user_ids)
+
+    invalid_users_ids = [
+        invalid_id
+        for invalid_id in user_ids
+        if invalid_id not in set(user_profiles.values_list("id", flat=True))
```

Comment:
You are recomputing this set on every iteration of the comprehension.

Reply:
ok will fix that, i assumed python should computed it only once

Reply:
If you're interested in working on performance projects like this, you'll want to make sure you understand this type of language semantics detail. There are plenty of good resources, both on the semantics of what a list comprehension does and how to write performant Python code; if you ever realize you are assuming something is fast that may or may not be, it's best to check.

Reply:
Sure, will keep that in mind